diff --git a/llvm/CMakeLists.txt b/llvm/CMakeLists.txt
index 3208147101c0..5fbac173aa1e 100644
--- a/llvm/CMakeLists.txt
+++ b/llvm/CMakeLists.txt
@@ -487,6 +487,7 @@ set(LLVM_ALL_TARGETS
   WebAssembly
   X86
   XCore
+  Cpu0
   )
 
 set(LLVM_ALL_EXPERIMENTAL_TARGETS
diff --git a/llvm/cmake/config-ix.cmake b/llvm/cmake/config-ix.cmake
index 0aae13e30f2a..d304dd0cf8bd 100644
--- a/llvm/cmake/config-ix.cmake
+++ b/llvm/cmake/config-ix.cmake
@@ -507,6 +507,8 @@ elseif (LLVM_NATIVE_ARCH STREQUAL "m68k")
   set(LLVM_NATIVE_ARCH M68k)
 elseif (LLVM_NATIVE_ARCH MATCHES "loongarch")
   set(LLVM_NATIVE_ARCH LoongArch)
+elseif (LLVM_NATIVE_ARCH MATCHES "cpu0")
+  set(LLVM_NATIVE_ARCH Cpu0)
 else ()
   message(FATAL_ERROR "Unknown architecture ${LLVM_NATIVE_ARCH}")
 endif ()
diff --git a/llvm/include/llvm/BinaryFormat/ELF.h b/llvm/include/llvm/BinaryFormat/ELF.h
index dfba18014991..0f6418af7cd6 100644
--- a/llvm/include/llvm/BinaryFormat/ELF.h
+++ b/llvm/include/llvm/BinaryFormat/ELF.h
@@ -320,6 +320,7 @@ enum {
   EM_VE = 251,            // NEC SX-Aurora VE
   EM_CSKY = 252,          // C-SKY 32-bit processor
   EM_LOONGARCH = 258,     // LoongArch
+  EM_CPU0 = 999           // CPU0
 };
 
 // Object file classes.
@@ -979,6 +980,19 @@ enum {
 #include "ELFRelocs/CSKY.def"
 };
 
+// Cpu0 Specific e_flags
+enum {
+  EF_CPU0_NOREORDER = 0x00000001, // Don't reorder instructions
+  EF_CPU0_PIC       = 0x00000002, // Position independent code
+  EF_CPU0_ARCH_32   = 0x50000000, // CPU032 instruction set per linux not elf.h
+  EF_CPU0_ARCH      = 0xf0000000  // Mask for applying EF_CPU0_ARCH_ variant
+};
+
+// ELF Relocation types for Mips
+enum {
+#include "ELFRelocs/Cpu0.def"
+};
+
 // LoongArch Specific e_flags
 enum : unsigned {
   // Definitions from LoongArch ELF psABI v2.01.
diff --git a/llvm/include/llvm/Object/ELFObjectFile.h b/llvm/include/llvm/Object/ELFObjectFile.h
index 8cc09e7fd7d5..7ddf49c8c99d 100644
--- a/llvm/include/llvm/Object/ELFObjectFile.h
+++ b/llvm/include/llvm/Object/ELFObjectFile.h
@@ -1271,6 +1271,8 @@ StringRef ELFObjectFile<ELFT>::getFileFormatName() const {
       return "elf32-loongarch";
     case ELF::EM_XTENSA:
       return "elf32-xtensa";
+    case ELF::EM_CPU0:        // 32bit ELF format name for CPU0
+      return "ELF32-cpu0";
     default:
       return "elf32-unknown";
     }
@@ -1403,6 +1405,8 @@ template <class ELFT> Triple::ArchType ELFObjectFile<ELFT>::getArch() const {
 
   case ELF::EM_XTENSA:
     return Triple::xtensa;
+  case ELF::EM_CPU0:        // get arch Triple for CPU0
+    return Triple::cpu0;
 
   default:
     return Triple::UnknownArch;
diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index b3bb354b38ff..27e84cb24cc2 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -65,6 +65,7 @@ public:
     mipsel,         // MIPSEL: mipsel, mipsallegrexe, mipsr6el
     mips64,         // MIPS64: mips64, mips64r6, mipsn32, mipsn32r6
     mips64el,       // MIPS64EL: mips64el, mips64r6el, mipsn32el, mipsn32r6el
+    cpu0,           // For Tutorial Backend Cpu0
     msp430,         // MSP430: msp430
     ppc,            // PPC: powerpc
     ppcle,          // PPCLE: powerpc (little endian)
diff --git a/llvm/lib/Object/ELF.cpp b/llvm/lib/Object/ELF.cpp
index 0ac4e7a57759..4b2c046de197 100644
--- a/llvm/lib/Object/ELF.cpp
+++ b/llvm/lib/Object/ELF.cpp
@@ -181,6 +181,13 @@ StringRef llvm::object::getELFRelocationTypeName(uint32_t Machine,
       break;
     }
     break;
+  case ELF::EM_CPU0:
+    switch (Type) {
+#include "llvm/BinaryFormat/ELFRelocs/Cpu0.def"
+    default:
+      break;
+    }
+    break;
   default:
     break;
   }
diff --git a/llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp b/llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp
index 0863345b0c6d..3b244ce326ea 100644
--- a/llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp
+++ b/llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp
@@ -28,7 +28,9 @@
 #include "llvm/Support/Casting.h"
 #include "llvm/Support/EndianStream.h"
 #include "llvm/Support/raw_ostream.h"
-
+#include <fstream>
+#include <sstream>
+#include <string>
 using namespace llvm;
 
 #define DEBUG_TYPE "mccodeemitter"
@@ -108,6 +110,8 @@ MCCodeEmitter *llvm::createRISCVMCCodeEmitter(const MCInstrInfo &MCII,
   return new RISCVMCCodeEmitter(Ctx, MCII);
 }
 
+std::map<std::string, std::string> param_map;
+
 // Expand PseudoCALL(Reg), PseudoTAIL and PseudoJump to AUIPC and JALR with
 // relocation types. We expand those pseudo-instructions while encoding them,
 // meaning AUIPC and JALR won't go through RISC-V MC to MC compressed
@@ -146,6 +150,44 @@ void RISCVMCCodeEmitter::expandFunctionCall(const MCInst &MI,
 
   const MCExpr *CallExpr = Func.getExpr();
 
+  if (MCExpr::Target == CallExpr->getKind()) {
+    const MCTargetExpr &TE = cast<MCTargetExpr>(*CallExpr);
+    const RISCVMCExpr &RE = cast<RISCVMCExpr>(TE);
+    if (MCExpr::SymbolRef == RE.getSubExpr()->getKind()) {
+      const MCSymbolRefExpr &SRE = cast<MCSymbolRefExpr>(*RE.getSubExpr());
+      StringRef Name = SRE.getSymbol().getName();
+      if (!Name.str().compare(0, 10, "run_on_npu", 0, 10)) {
+        if (param_map.size() == 0) {
+          std::ifstream file("param_map.txt");
+          if (file.is_open()) {
+            std::string line;
+            while (getline(file, line)) {
+              std::istringstream tokenStream(line);
+              std::string part;
+              std::vector<std::string> parts;
+              while (getline(tokenStream, part, '|')) {
+                parts.push_back(part);
+              }
+              if (parts.size() != 3) {
+                printf("Invalid line in param_map.txt\n");
+                exit(1);
+              }
+              param_map[parts[0]] = parts[2];
+            }
+          }
+        }
+        if (param_map.count(Name.str())) {
+          std::string bc = param_map[Name.str()];
+          std::istringstream iss(bc);
+          while (iss >> Binary) {
+            support::endian::write(CB, Binary, llvm::endianness::little);
+          }
+          printf("handle func: %s\n", Name.str().c_str());
+          return;
+        }
+      }
+    }
+  }
   // Emit AUIPC Ra, Func with R_RISCV_CALL relocation type.
   TmpInst = MCInstBuilder(RISCV::AUIPC).addReg(Ra).addExpr(CallExpr);
   Binary = getBinaryCodeForInstr(TmpInst, Fixups, STI);
diff --git a/llvm/lib/TargetParser/Triple.cpp b/llvm/lib/TargetParser/Triple.cpp
index 4fc1ff5aaa05..9a8c696f9ed0 100644
--- a/llvm/lib/TargetParser/Triple.cpp
+++ b/llvm/lib/TargetParser/Triple.cpp
@@ -70,6 +70,7 @@ StringRef Triple::getArchTypeName(ArchType Kind) {
   case sparcel:        return "sparcel";
   case sparcv9:        return "sparcv9";
   case spir64:         return "spir64";
+  case cpu0:           return "cpu0"; // Arch type name for CPU0
   case spir:           return "spir";
   case spirv:          return "spirv";
   case spirv32:        return "spirv32";
@@ -189,7 +190,7 @@ StringRef Triple::getArchTypePrefix(ArchType Kind) {
   case sparc:       return "sparc";
 
   case systemz:     return "s390";
-
+  case cpu0:        return "cpu0"; // Arch type prefix for CPU0
   case x86:
   case x86_64:      return "x86";
 
@@ -452,6 +453,7 @@ Triple::ArchType Triple::getArchTypeForLLVMName(StringRef Name) {
     .Case("renderscript64", renderscript64)
     .Case("ve", ve)
     .Case("csky", csky)
+    .Case("cpu0", cpu0)         // Arch type for LLVM name for CPU0
     .Case("loongarch32", loongarch32)
     .Case("loongarch64", loongarch64)
     .Case("dxil", dxil)
@@ -598,6 +600,7 @@ static Triple::ArchType parseArch(StringRef ArchName) {
           .Case("wasm32", Triple::wasm32)
           .Case("wasm64", Triple::wasm64)
           .Case("csky", Triple::csky)
+          .Case("cpu0", Triple::cpu0) // string to triple for CPU0
           .Case("loongarch32", Triple::loongarch32)
           .Case("loongarch64", Triple::loongarch64)
           .Cases("dxil", "dxilv1.0", "dxilv1.1", "dxilv1.2", "dxilv1.3",
@@ -903,6 +906,7 @@ static Triple::ObjectFormatType getDefaultFormat(const Triple &T) {
   case Triple::hexagon:
   case Triple::hsail64:
   case Triple::hsail:
+  case Triple::cpu0:        // Default format to CPU0 is ELF
   case Triple::kalimba:
   case Triple::lanai:
   case Triple::le32:
@@ -1625,6 +1629,7 @@ unsigned Triple::getArchPointerBitWidth(llvm::Triple::ArchType Arch) {
   case llvm::Triple::thumbeb:
   case llvm::Triple::wasm32:
   case llvm::Triple::x86:
+  case llvm::Triple::cpu0:         // CPU0's pointer width is 32 bits.
   case llvm::Triple::xcore:
   case llvm::Triple::xtensa:
     return 32;
@@ -1708,6 +1713,7 @@ Triple Triple::get32BitArchVariant() const {
   case Triple::riscv32:
   case Triple::shave:
   case Triple::sparc:
+  case Triple::cpu0:        // CPU0 is 32-bit.
   case Triple::sparcel:
   case Triple::spir:
   case Triple::spirv32:
@@ -1765,6 +1771,7 @@ Triple Triple::get64BitArchVariant() const {
   case Triple::m68k:
   case Triple::msp430:
   case Triple::r600:
+  case Triple::cpu0:  // cpu0 has not 64-bit arch.
   case Triple::shave:
   case Triple::sparcel:
   case Triple::tce:
@@ -1874,6 +1881,7 @@ Triple Triple::getBigEndianArchVariant() const {
   case Triple::ve:
   case Triple::csky:
   case Triple::xtensa:
+  case Triple::cpu0:    // CPU0 is not big endian.
 
   // ARM is intentionally unsupported here, changing the architecture would
   // drop any arch suffixes.
@@ -1985,6 +1993,7 @@ bool Triple::isLittleEndian() const {
   case Triple::x86_64:
   case Triple::xcore:
   case Triple::xtensa:
+  case Triple::cpu0:    // CPU0 is little endia
     return true;
   default:
     return false;
